src/components/implementation/no_interface/vkernel/vk_api.c

```diff
@@ -1,71 +1,105 @@
 #include "vk_api.h"
 
+extern vaddr_t cos_upcall_entry;
+
+static struct cos_aep_info *
+vm_schedaep_get(struct vms_info *vminfo)
+{ return cos_sched_aep_get(&(vminfo->dci)); }
+
 void
 vk_initcaps_init(struct vms_info *vminfo, struct vkernel_info *vkinfo)
 {
-	struct cos_compinfo *vmcinfo = &vminfo->cinfo;
-	struct cos_compinfo *vkcinfo = &vkinfo->cinfo;
+	struct cos_compinfo *vk_cinfo = cos_compinfo_get(cos_defcompinfo_curr_get());
+	struct cos_defcompinfo *vmdci = &(vminfo->dci);
+	struct cos_compinfo *vmcinfo  = cos_compinfo_get(vmdci);
+	struct cos_aep_info *initaep  = cos_sched_aep_get(vmdci);
+	pgtblcap_t vmutpt;
 	int ret;
 
 	assert(vminfo && vkinfo);	
 
-	vminfo->exitthd = cos_thd_alloc(vkcinfo, vkcinfo->comp_cap, vm_exit, (void *)vminfo->id);
-	assert(vminfo->exitthd);
+	vmutpt = cos_pgtbl_alloc(vk_cinfo);
+	assert(vmutpt);
+
+	cos_meminfo_init(&(vmcinfo->mi), BOOT_MEM_KM_BASE, VM_UNTYPED_SIZE, vmutpt);
+	ret = cos_defcompinfo_child_alloc(vmdci, (vaddr_t)&cos_upcall_entry, (vaddr_t)BOOT_MEM_VM_BASE,
+					  VM_CAPTBL_FREE, 1);
+	cos_compinfo_init(&(vminfo->shm_cinfo), vmcinfo->pgtbl_cap, vmcinfo->captbl_cap, vmcinfo->comp_cap,
+			  (vaddr_t)VK_VM_SHM_BASE, VM_CAPTBL_FREE, vk_cinfo);
+
+	printc("\tCreating and copying initial component capabilities\n");
+	ret = cos_cap_cpy_at(vmcinfo, BOOT_CAPTBL_SELF_CT, vk_cinfo, vmcinfo->captbl_cap);
+	assert(ret == 0);
+	ret = cos_cap_cpy_at(vmcinfo, BOOT_CAPTBL_SELF_PT, vk_cinfo, vmcinfo->pgtbl_cap);
+	assert(ret == 0);
+	ret = cos_cap_cpy_at(vmcinfo, BOOT_CAPTBL_SELF_UNTYPED_PT, vk_cinfo, vmutpt);
+	assert(ret == 0);
+	ret = cos_cap_cpy_at(vmcinfo, BOOT_CAPTBL_SELF_COMP, vk_cinfo, vmcinfo->comp_cap);
+	assert(ret == 0);
 
-	vminfo->initthd = cos_thd_alloc(vkcinfo, vmcinfo->comp_cap, vm_init, (void *)vminfo->id);
-	assert(vminfo->initthd);
-	vminfo->inittid = (thdid_t)cos_introspect(vkcinfo, vminfo->initthd, THD_GET_TID);
-	printc("\tInit thread= cap:%x tid:%x\n", (unsigned int)vminfo->initthd, (unsigned int)vminfo->inittid);
+	printc("\tInit thread= cap:%x\n", (unsigned int)initaep->thd);
```

Comment:
If it sin't a library, I'm with you.  If it is a library (and I'd like it to be), then it isn't desirable.